File: scripts/generate_alarm_sounds.py
#!/usr/bin/env python3
"""Generate lightweight alarm sound WAV files for SmartWake."""

import math
import struct
import wave
from pathlib import Path

SAMPLE_RATE = 44100
OUTPUT_DIR = Path(__file__).resolve().parent.parent / "assets" / "sounds"


def write_wav(path: Path, samples: list[float], sample_rate: int = SAMPLE_RATE) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with wave.open(str(path), "w") as wf:
        wf.setnchannels(1)
        wf.setsampwidth(2)
        wf.setframerate(sample_rate)
        frames = bytearray()
        for sample in samples:
            clamped = max(-1.0, min(1.0, sample))
            frames.extend(struct.pack("<h", int(clamped * 32767)))
        wf.writeframes(frames)


def sine_tone(freq: float, duration: float, volume: float = 0.35) -> list[float]:
    total = int(SAMPLE_RATE * duration)
    return [
        volume * math.sin(2 * math.pi * freq * i / SAMPLE_RATE)
        for i in range(total)
    ]


def envelope(samples: list[float], attack: float = 0.02, release: float = 0.15) -> list[float]:
    attack_samples = int(SAMPLE_RATE * attack)
    release_samples = int(SAMPLE_RATE * release)
    out = samples[:]
    for i in range(min(attack_samples, len(out))):
        out[i] *= i / max(1, attack_samples)
    for i in range(min(release_samples, len(out))):
        idx = len(out) - 1 - i
        out[idx] *= i / max(1, release_samples)
    return out


def silence(duration: float) -> list[float]:
    return [0.0] * int(SAMPLE_RATE * duration)


def gentle_chime() -> list[float]:
    notes = [523.25, 659.25, 783.99]
    audio: list[float] = []
    for freq in notes:
        audio.extend(envelope(sine_tone(freq, 0.55, 0.28)))
        audio.extend(silence(0.08))
    return audio


def morning_birds() -> list[float]:
    audio: list[float] = []
    pattern = [(880, 0.12), (1175, 0.1), (988, 0.14), (1319, 0.11)]
    for _ in range(6):
        for freq, duration in pattern:
            chirp = envelope(sine_tone(freq, duration, 0.22))
            audio.extend(chirp)
            audio.extend(silence(0.05))
        audio.extend(silence(0.2))
    return audio


def soft_piano() -> list[float]:
    notes = [261.63, 329.63, 392.0, 523.25]
    audio: list[float] = []
    for freq in notes:
        tone = sine_tone(freq, 0.45, 0.24)
        harmonic = sine_tone(freq * 2, 0.45, 0.08)
        mixed = [a + b for a, b in zip(tone, harmonic)]
        audio.extend(envelope(mixed))
        audio.extend(silence(0.06))
    return audio


def ocean_waves() -> list[float]:
    duration = 4.0
    total = int(SAMPLE_RATE * duration)
    audio: list[float] = []
    for i in range(total):
        t = i / SAMPLE_RATE
        low = math.sin(2 * math.pi * 0.25 * t)
        mid = math.sin(2 * math.pi * 0.6 * t + math.sin(t))
        noise = math.sin(2 * math.pi * (3 + (i % 17) / 17) * t) * 0.08
        sample = 0.18 * low + 0.12 * mid + noise
        audio.append(sample)
    return envelope(audio, attack=0.4, release=0.4)


def sunrise() -> list[float]:
    duration = 3.5
    total = int(SAMPLE_RATE * duration)
    audio: list[float] = []
    for i in range(total):
        t = i / SAMPLE_RATE
        progress = t / duration
        freq = 220 + progress * 440
        sample = 0.3 * math.sin(2 * math.pi * freq * t)
        audio.append(sample)
    return envelope(audio, attack=0.5, release=0.25)


SOUNDS = {
    "gentle_chime": gentle_chime,
    "morning_birds": morning_birds,
    "soft_piano": soft_piano,
    "ocean_waves": ocean_waves,
    "sunrise": sunrise,
}


def main() -> None:
    for name, builder in SOUNDS.items():
        path = OUTPUT_DIR / f"{name}.wav"
        write_wav(path, builder())
        print(f"Wrote {path}")


if __name__ == "__main__":
    main()

File: scripts/generate_alarm_sounds.dart
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

/// Generates alarm WAV files. Run: dart run scripts/generate_alarm_sounds.dart
void main() {
  const sampleRate = 44100;
  final outDir = Directory('assets/sounds');
  outDir.createSync(recursive: true);

  final sounds = <String, List<double> Function()>{
    'gentle_chime': _gentleChime,
    'morning_birds': _morningBirds,
    'soft_piano': _softPiano,
    'ocean_waves': _oceanWaves,
    'sunrise': _sunrise,
  };

  for (final entry in sounds.entries) {
    final path = File('${outDir.path}/${entry.key}.wav');
    _writeWav(path, entry.value(), sampleRate);
    stdout.writeln('Wrote ${path.path}');
  }
}

void _writeWav(File path, List<double> samples, int sampleRate) {
  final data = BytesBuilder();
  for (final sample in samples) {
    final clamped = sample.clamp(-1.0, 1.0);
    final int16 = (clamped * 32767).round().clamp(-32768, 32767);
    data.addByte(int16 & 0xFF);
    data.addByte((int16 >> 8) & 0xFF);
  }
  final pcm = data.toBytes();
  final header = _wavHeader(pcm.length, sampleRate);
  path.writeAsBytesSync([...header, ...pcm]);
}

List<int> _wavHeader(int dataLength, int sampleRate) {
  final byteRate = sampleRate * 2;
  final blockAlign = 2;
  final chunkSize = 36 + dataLength;
  final buffer = BytesBuilder();
  void writeString(String s) => buffer.add(s.codeUnits);
  void writeInt32(int v) => buffer.add([v & 0xFF, (v >> 8) & 0xFF, (v >> 16) & 0xFF, (v >> 24) & 0xFF]);
  void writeInt16(int v) => buffer.add([v & 0xFF, (v >> 8) & 0xFF]);

  writeString('RIFF');
  writeInt32(chunkSize);
  writeString('WAVE');
  writeString('fmt ');
  writeInt32(16);
  writeInt16(1);
  writeInt16(1);
  writeInt32(sampleRate);
  writeInt32(byteRate);
  writeInt16(blockAlign);
  writeInt16(16);
  writeString('data');
  writeInt32(dataLength);
  return buffer.toBytes();
}

List<double> _sine(double freq, double duration, {double volume = 0.35, int sampleRate = 44100}) {
  final total = (sampleRate * duration).round();
  return List.generate(total, (i) => volume * sin(2 * pi * freq * i / sampleRate));
}

List<double> _envelope(List<double> samples, {double attack = 0.02, double release = 0.15, int sampleRate = 44100}) {
  final out = List<double>.from(samples);
  final attackSamples = (sampleRate * attack).round();
  final releaseSamples = (sampleRate * release).round();
  for (var i = 0; i < min(attackSamples, out.length); i++) {
    out[i] *= i / max(1, attackSamples);
  }
  for (var i = 0; i < min(releaseSamples, out.length); i++) {
    final idx = out.length - 1 - i;
    out[idx] *= i / max(1, releaseSamples);
  }
  return out;
}

List<double> _silence(double duration, {int sampleRate = 44100}) =>
    List.filled((sampleRate * duration).round(), 0.0);

List<double> _gentleChime() {
  final audio = <double>[];
  for (final freq in [523.25, 659.25, 783.99]) {
    audio.addAll(_envelope(_sine(freq, 0.55, volume: 0.28)));
    audio.addAll(_silence(0.08));
  }
  return audio;
}

List<double> _morningBirds() {
  final audio = <double>[];
  const pattern = [(880.0, 0.12), (1175.0, 0.1), (988.0, 0.14), (1319.0, 0.11)];
  for (var r = 0; r < 6; r++) {
    for (final (freq, duration) in pattern) {
      audio.addAll(_envelope(_sine(freq, duration, volume: 0.22)));
      audio.addAll(_silence(0.05));
    }
    audio.addAll(_silence(0.2));
  }
  return audio;
}

List<double> _softPiano() {
  final audio = <double>[];
  for (final freq in [261.63, 329.63, 392.0, 523.25]) {
    final tone = _sine(freq, 0.45, volume: 0.24);
    final harmonic = _sine(freq * 2, 0.45, volume: 0.08);
    final mixed = List<double>.generate(tone.length, (i) => tone[i] + harmonic[i]);
    audio.addAll(_envelope(mixed));
    audio.addAll(_silence(0.06));
  }
  return audio;
}

List<double> _oceanWaves() {
  const duration = 4.0;
  const sampleRate = 44100;
  final total = (sampleRate * duration).round();
  return _envelope(List.generate(total, (i) {
    final t = i / sampleRate;
    final low = sin(2 * pi * 0.25 * t);
    final mid = sin(2 * pi * 0.6 * t + sin(t));
    final noise = sin(2 * pi * (3 + (i % 17) / 17) * t) * 0.08;
    return 0.18 * low + 0.12 * mid + noise;
  }), attack: 0.4, release: 0.4);
}

List<double> _sunrise() {
  const duration = 3.5;
  const sampleRate = 44100;
  final total = (sampleRate * duration).round();
  return _envelope(List.generate(total, (i) {
    final t = i / sampleRate;
    final progress = t / duration;
    final freq = 220 + progress * 440;
    return 0.3 * sin(2 * pi * freq * t);
  }), attack: 0.5, release: 0.25);
}
